### PR TITLE
feat(wallet): Add hashes to Balance/AssetBalance

### PR DESCRIPTION
**Pull request by endkeyCoder**
Hi @snowypowers I followed the guidelines you gave me, as soon as you can, check if you agree with the implementation, thanks for the help :)

### CODE REVIEW

File: packages/neon-core/src/wallet/Balance.ts [hunk 1]
```diff
@@ -100,6 +101,8 @@ export class Balance {
   /** The token balances in this Balance for each token keyed by its symbol. */
   public tokens: { [sym: string]: Fixed8 };
 
+  public tokenHashes: { [sym: string]: string };
```
Comment (snowypowers) [inline packages/neon-core/src/wallet/Balance.ts:104]: At the moment it looks like the hash is not guaranteed to exists even if the token is present within the balance. Is it possible to guarantee it?

File: packages/neon-api/__tests__/provider/neoCli/core.ts [hunk 2]
```diff
@@ -136,7 +136,7 @@ describe("getClaims", () => {
   test("returns successful claims", async () => {
     const testClaim = {
       end_height: 231,
-      generated: 1464,
+      generated: 0,
```
Comment (snowypowers) [inline packages/neon-api/__tests__/provider/neoCli/core.ts:139]: Please use a non-zero unique number for `generated` and `sys_fee` since the code defaults to `0`.

File: packages/neon-api/__tests__/provider/neoscan/core.ts [hunk 3]
```diff
@@ -118,6 +118,7 @@ describe("getBalance", () => {
         tokens: {
           TEST: 100,
         },
+        tokenHashes: {},
```
Comment (snowypowers) [inline packages/neon-api/__tests__/provider/neoscan/core.ts:121]: Is this possible? Will Neoscan ever return a filled token balance with no hash?

File: packages/neon-api/src/provider/neoscan/core.ts [hunk 4]
```diff
@@ -120,10 +122,14 @@ export async function getBalance(
     if (b.amount > 0 && b.unspent.length > 0) {
       bal.addAsset(b.asset, {
         unspent: parseUnspent(b.unspent),
+        hash: b.asset_hash,
       } as Partial<wallet.AssetBalanceLike>);
     } else {
       bal.addToken(b.asset, b.amount);
     }
+    if (b.asset && b.asset_hash) {
```
Comment (snowypowers) [inline packages/neon-api/src/provider/neoscan/core.ts:130]: This means that we are adding a token hash even if the logic is adding an asset instead. I do not think that is the correct case.

File: packages/neon-core/src/wallet/components/AssetBalance.ts [hunk 5]
```diff
@@ -29,6 +31,8 @@ export class AssetBalance implements NeonObject {
     this.unconfirmed = abLike.unconfirmed
       ? abLike.unconfirmed.map((coin) => new Coin(coin))
       : [];
+
+    this.hash = abLike.hash || "";
```
Comment (snowypowers) [inline packages/neon-core/src/wallet/components/AssetBalance.ts:35]: Once again, should we be allowing this field to be unpopulated? I would say no unless we have serious issues trying to fill this field.

**Review by snowypowers — CHANGES_REQUESTED** [review]
So far i have not seen a test case where the hash is actually populated. Please exercise the code correctly in tests.

File: packages/neon-core/src/wallet/Balance.ts (hunk 1 above)
Comment (endkeyCoder) [reply packages/neon-core/src/wallet/Balance.ts:104]: when querying the get_balance endpoint in both neoscan and dora, the hash is always present, the "amount" field being greater or equal than zero, and it is also present when the unspent array size is zero or greater than zero. Wouldn't that be the way to guarantee that?

File: packages/neon-api/__tests__/provider/neoCli/core.ts (hunk 2 above)
Comment (endkeyCoder) [reply packages/neon-api/__tests__/provider/neoCli/core.ts]: correction applied

File: packages/neon-api/__tests__/provider/neoscan/core.ts (hunk 3 above)
Comment (endkeyCoder) [reply packages/neon-api/__tests__/provider/neoscan/core.ts]: correction applied

File: packages/neon-api/src/provider/neoscan/core.ts (hunk 4 above)
Comment (endkeyCoder) [reply packages/neon-api/src/provider/neoscan/core.ts]: correction applied

File: packages/neon-core/src/wallet/components/AssetBalance.ts (hunk 5 above)
Comment (endkeyCoder) [reply packages/neon-core/src/wallet/components/AssetBalance.ts]: correction applied

**Comment by endkeyCoder** [comment]
I followed the guidelines and applied some corrections, I await further feedback, thank you @snowypowers :)

File: packages/neon-core/src/wallet/components/AssetBalance.ts [hunk 6]
```diff
@@ -20,15 +24,18 @@ export class AssetBalance implements NeonObject {
   public unspent: Coin[];
   public spent: Coin[];
   public unconfirmed: Coin[];
+  public hash: string;
 
-  public constructor(abLike: Partial<AssetBalanceLike> = {}) {
+  public constructor(abLike: AssetBalanceWithHash = { hash: "nonHash" }) {
```
Comment (snowypowers) [inline packages/neon-core/src/wallet/components/AssetBalance.ts:29]: This kinds of defeats the purpose of the type system. If we are filling in a non-hash string for the user, it just means that the field itself is not guaranteed. Why not just use the old Partial<AssetBalanceLike> and throw an error if the hash field is empty.

Comment (endkeyCoder) [reply packages/neon-core/src/wallet/components/AssetBalance.ts:29]: it a good suggestion, i will do this

**Comment by endkeyCoder** [comment]
I made the necessary changes, I await your review @snowypowers tks :)

File: packages/neon-core/__tests__/tx/calculate.ts [hunk 7]
```diff
@@ -42,12 +48,15 @@ describe("calculateInputs", () => {
...
           unspent: [{ value: 10, txid: "ab", index: 0 }],
+          hash: NEO_ASSET_HASH,
         },
         GAS: {
           unspent: [{ value: 1.5, txid: "cd", index: 0 }],
+          hash: GAS_ASSET_HASH,
         },
         TEST: {
           unspent: [{ value: 1, txid: "ef", index: 0 }],
+          hash: NEO_ASSET_HASH,
```
Comment (snowypowers) [inline packages/neon-core/__tests__/tx/calculate.ts:59]: Why does the TEST asset have the the NEO_ASSET_HASH?

File: packages/neon-core/__tests__/wallet/Balance.ts [hunk 8]
```diff
@@ -23,9 +26,10 @@ describe("constructor", () => {
       assetSymbols: ["ASS"],
       tokenSymbols: ["TOK"],
       assets: {
-        ASS: {} as AssetBalanceLike,
+        ASS: { hash: NEO_ASSET_HASH } as AssetBalanceLike,
```
Comment (snowypowers) [inline packages/neon-core/__tests__/wallet/Balance.ts:29]: Can we not use the name NEO_ASSET_HASH if it is not? This is confusing.

File: packages/neon-core/__tests__/wallet/Balance.ts [hunk 9]
```diff
@@ -66,11 +71,13 @@ describe("export", () => {
     const expected = {
       address: "address",
       net: "UnitTestNet",
+      tokenHashes: {},
```
Comment (snowypowers) [inline packages/neon-core/__tests__/wallet/Balance.ts:74]: Since there is the TOK token, I expect a hash for TOK here.

File: packages/neon-core/__tests__/wallet/components/AssetBalance.ts [hunk 10]
```diff
@@ -1,15 +1,22 @@
-import AssetBalance from "../../../src/wallet/components/AssetBalance";
+import AssetBalance, {
+  EXCEPTION_HASH_MSG,
```
Comment (snowypowers) [inline packages/neon-core/__tests__/wallet/components/AssetBalance.ts:2]: What is this import for?

File: packages/neon-core/src/wallet/verify.ts [hunk 11]
```diff
@@ -138,3 +138,9 @@ export function isAddress(address: string): boolean {
     return false;
   }
 }
+
+export function isAssetHash(assetHash: string): boolean {
+  return (
+    (isHex(assetHash) && assetHash.length === 40) || assetHash.length === 64
```
Comment (snowypowers) [inline packages/neon-core/src/wallet/verify.ts:144]: I think you are mixing up assets and tokens here. Assets are UTXO styled with the 64 char hash. The 40 char hash belongs to tokens.

File: packages/neon-core/src/wallet/components/AssetBalance.ts [hunk 12]
```diff
@@ -29,6 +34,11 @@ export class AssetBalance implements NeonObject {
     this.unconfirmed = abLike.unconfirmed
       ? abLike.unconfirmed.map((coin) => new Coin(coin))
       : [];
+
+    if (!abLike.hash || !isAssetHash(abLike.hash)) {
+      throw new Error(EXCEPTION_HASH_MSG);
```
Comment (snowypowers) [inline packages/neon-core/src/wallet/components/AssetBalance.ts:39]: Please inline the message. We do not need to export error strings.

File: packages/neon-core/__integration__/rpc/RPCClient.ts [hunk 13]
```diff
@@ -12,7 +12,7 @@ const TESTNET_URLS = [
 ];
 
 let client: rpc.RPCClient;
-const address = "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW";
+const address = "AXSoNQEKjmqPBNPg5cNrHyWivfjok3Vj9D";
```
Comment (snowypowers) [inline packages/neon-core/__integration__/rpc/RPCClient.ts:15]: Is this change necessary?

Comment (endkeyCoder) [reply packages/neon-core/__integration__/rpc/RPCClient.ts:15]: yes, because the other address doesn't have a balance

File: packages/neon-core/src/wallet/verify.ts (hunk 11 above)
Comment (endkeyCoder) [reply packages/neon-core/src/wallet/verify.ts:144]: ah ok, so just a "isHex" and "assetHash.lenght === 64" validation are necessary?

File: packages/neon-core/__tests__/tx/calculate.ts (hunk 7 above)
Comment (endkeyCoder) [reply packages/neon-core/__tests__/tx/calculate.ts:59]: we're going to need a valid hash to test, I thought I'd use the NEO hash as a standard hash, but it could be another one

Comment (snowypowers) [reply packages/neon-core/__tests__/tx/calculate.ts:59]: This is a unit test so just generate a random hash would do.

File: packages/neon-core/src/wallet/verify.ts (hunk 11 above)
Comment (snowypowers) [reply packages/neon-core/src/wallet/verify.ts:144]: yups i believe those are the only conditions.

Comment (endkeyCoder) [reply packages/neon-core/src/wallet/verify.ts:144]: ok but i checked the get_balance answer with other tokens like RPX or PHX and the asset_hash of those tokens is 40 characters long, wouldn't that cause a problem?

File: packages/neon-core/__integration__/rpc/RPCClient.ts (hunk 13 above)
Comment (snowypowers) [reply packages/neon-core/__integration__/rpc/RPCClient.ts:15]: What balance are you looking at? So I mainly use that address since the private key is available in the repository for use. It would be simpler if we can continue to use the test keys instead of a custom address that the repository does not control for integration tests.

File: packages/neon-core/src/wallet/verify.ts (hunk 11 above)
Comment (snowypowers) [reply packages/neon-core/src/wallet/verify.ts:144]: I think there is a misconception here. RPX and PHX are not assets. They are NEP5 tokens and do not use UTXO. Only GAS and NEO are assets so far in neo2.

File: packages/neon-core/__tests__/tx/calculate.ts (hunk 7 above)
Comment (endkeyCoder) [reply packages/neon-core/__tests__/tx/calculate.ts:59]: ok, I will do it

File: packages/neon-core/src/wallet/components/AssetBalance.ts (hunk 12 above)
Comment (endkeyCoder) [reply packages/neon-core/src/wallet/components/AssetBalance.ts:39]: ok, I will do it

File: packages/neon-core/src/wallet/verify.ts (hunk 11 above)
Comment (endkeyCoder) [reply packages/neon-core/src/wallet/verify.ts:144]: ok i really didn't notice it, I will fix the function, tks

File: packages/neon-core/__integration__/rpc/RPCClient.ts (hunk 13 above)
Comment (endkeyCoder) [reply packages/neon-core/__integration__/rpc/RPCClient.ts:15]: I really don't need to change the address, I'll go back to the other address

**Comment by endkeyCoder** [comment]
applied corrections @snowypowers thank you by your help :)

File: packages/neon-core/__tests__/wallet/components/AssetBalance.ts (hunk 10 above)
Comment (endkeyCoder) [reply packages/neon-core/__tests__/wallet/components/AssetBalance.ts]: it was a mistake, it's fixed

File: packages/neon-core/__tests__/wallet/Balance.ts (hunk 8 above)
Comment (endkeyCoder) [reply packages/neon-core/__tests__/wallet/Balance.ts]: I agree, I changed it to DEFAULT_HASH

**Review by snowypowers — APPROVED** [review]

**Comment by snowypowers** [comment]
please update the PR description and title to describe the change that is useful as a standalone commit.

**Comment by endkeyCoder** [comment]
Below is an example of the function's return format, new fields were added to make the information more complete
```
getClaims: {
      end_height: 231,
      generated: 1337,
      n: 0,
      start_height: 48,
      sys_fee: 1337,
      txid: "dc44739e2f97743f2ed258988327560e2185ed13eec0097938eef4aea584bf04",
      unclaimed: 1464,
      value: 100000000,
    }

getBalance: {
        "net": "someUrl",
        "address":"someAddress",
        "assetSymbols": ["GAS", "NEO"],
        "assets": {
          "NEO": {
            "unspent": [{ "value": 2, "txid": "1", "index": 1 }],
            "balance": 2,
            "hash": "c56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b",
          },
          "GAS": {
            "unspent": [{ "value": 5, "txid": "1", "index": 1 }],
            "balance": 5,
            "hash": "602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7",
          },
        },
        "tokenSymbols": ["NEO"],
        "tokens": {
          "NEO": 21,
        },
        "tokenHashes": { NEO: "c56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b" },
    }
```
@snowypowers what do you think of this description?

**Comment by snowypowers** [comment]
You dont have to put up an example in the description, that should be left to the tests. What I will usually put is the main changes to the API. In this PR, the core change is actually adding the tokenHashes/hash fields to the Balance/AssetBalance class. These are driven by the data available from the API.

**Comment by endkeyCoder** [comment]
Added the "hash" / "tokenHashes" fields to the "Balance" / "Asset Balance" class according to the data available in the API
Fix #742

**Comment by endkeyCoder** [comment]
@snowypowers that way is good?

**Comment by endkeyCoder** [comment]
Hi @snowypowers , can we do merge of the code or lack something yet?
